src/api/app/models/staging_project.rb

```diff
@@ -0,0 +1,142 @@
+  def requests_to_review
+    @requests_to_review ||= BsRequest.with_open_reviews_for(by_project: name)
+  end
+
+  def building_repositories
+    set_buildinfo unless @building_repositories
+    @building_repositories
+  end
+
+  def broken_packages
+    set_buildinfo unless @broken_packages
+    @broken_packages
+  end
+
+  def missing_reviews
+    return @missing_reviews if @missing_reviews
+
+    @missing_reviews = []
+    attribs = [:by_group, :by_user, :by_project, :by_package]
+
+    (requests_to_review + staged_requests).uniq.each do |request|
+      request.reviews.each do |review|
+        next if review.state.to_s == 'accepted' || review.by_project == name
+        # FIXME: this loop (and the inner if) would not be needed
+        # if every review only has one valid by_xxx.
+        # I'm keeping it to mimic the python implementation.
+        # Instead, we could have something like
+        # who = review.by_group || review.by_user || review.by_project || review.by_package
+        attribs.each do |att|
+          who = review.send(att)
+          if who
+            @missing_reviews << { id: review.id, request: request.number, state: review.state.to_s, package: request.first_target_package, by: who }
+          end
+        end
+      end
+    end
+
+    @missing_reviews
+  end
+
+  def overall_state
+    @overall_state ||= state
+  end
+
+  private
+
+  def state
+    return :empty unless staged_requests.exists?
+    return :unacceptable if untracked_requests.present? || staged_requests.obsolete.present?
+
+    overall_state = build_state
+    overall_state = check_state if overall_state == :acceptable
+
+    return :review if overall_state == :acceptable && missing_reviews.present?
+
+    overall_state
+  end
+
+  def build_state
+    set_buildinfo
+
+    return :building if building_repositories.present?
+    return :failed if broken_packages.present?
+
+    :acceptable
+  end
+
+  def relevant_status_reports
+    @relevant_status_reports ||= status_reports.where(uuid: repositories.map(&:build_id))
+  end
+
+  def missing_checks?
+    relevant_status_reports.any? { |report| report.missing_checks.present? }
+  end
+
+  def check_state
+    status_checks = Status::Check.where(status_reports_id: relevant_status_reports)
+    return :testing if missing_checks? || status_checks.pending.exists?
+    return :failed if status_checks.failed.exists?
+    return :acceptable
+  end
+
+  def set_buildinfo
+    buildresult = Buildresult.find_hashed(project: name, view: 'summary')
+
+    # data of each pkg/repo/arch combo with state broken or failed
+    @broken_packages = []
+    # data and count of all repos that are currently building
+    @building_repositories = []
+
+    buildresult.elements('result') do |result|
+      building = ['published', 'unpublished'].exclude?(result['state']) || result['dirty'] == 'true'
+
+      result.elements('status') do |status|
+        code = status.get('code')
+
```

Comment:
[nitpick] I would remove this empty line, there are not conceptually different code blocks